Adds field selection RFC (#35)
graphql/composite-schemas-spec: spec/Appendix A -- Field Selection.md

in the `@is` and `@requires` directives.

To illustrate, consider a simple example from a GraphQL schema:

```graphql
type Query {
    userById(userId: ID! @is(field: "id")): User! @lookup
}
```

In this schema, the `userById` query uses the `@is` directive with {FieldSelectionMap} to declare that
the `userId` argument is semantically equivalent to the `User.id` field. 

An example query might look like this:

```graphql
query {
    userById(userId: "123") {
        id
    }
}
```

Here, it is exptected that the `userId` "123" corresponds directly to `User.id`, resulting in the
following response if correctly implemented:

```json
{
    "data": {
        "userById": {
            "id": "123"
        }
    }
}
```

The {FieldSelectionMap} scalar type is used to establish semantic equivalence between an argument and
the fields within the associated return type. To accomplish this, the scalar must define the
relationship between input fields or arguments and the fields in the resulting object. 
Consequently, a {FieldSelectionMap} only makes sense in the context of a specific argument and its 

> return type.

Reviewer: I think the situation is different for `@require`. When you put `@require` on an argument, the mapping you're describing is not between the argument and the *return type* of the field, but between the argument and the `object type` the field is defined on. Take this example:

```graphql
type Product {
  id: ID!
  delivery(
    zip: String!
    size: Int! @require(field: "dimension { size }")
    weight: Int! @require(field: "dimension { weight }")
  ): DeliveryEstimates
}
```

In here, `"dimension { size }"` refers to a field of `Product`, not `DeliveryEstimates `.

I think this means we probably need to describe this in more neutral terms that apply both to the `@is` and the `@require` use cases. Perhaps we could state that a `{FieldSelectionMap}` always describes a mapping to a composite type, but that directives differ in how they establish that context.

Author: good catch, resolved

The {FieldSelectionMap} scalar is represented as a string that, when parsed, produces a {SelectedValue}.

A {SelectedValue} must exactly match the shape of the argument value to be considered
valid. For non-scalar arguments, you must specify each field of the input type in
{SelectedObjectValue}.

```graphql example
extend type Query {
    findUserByName(user: UserInput! @is(field: "{ firstName: firstName }")): User @lookup
}
```

```graphql counter-example
extend type Query {

>     findUserByName(user: UserInput! @is(field: "firstName")): User @lookup

Reviewer: I agree this example is invalid, because the result is a scalar and not an object value. But I still think we should allow a shorthand when the field names match, so `{ firstName }` *would* be valid (similar to how JavaScript allows shorthand property definitions in object literals).

}
```

## Language

According to the GraphQL specification, an argument is a key-value pair in which the key is the name
of the argument and the value is a `Value`.

The `Value` of an argument can take various forms: it might be a scalar value (such as `Int`,
`Float`, `String`, `Boolean`, `Null`, or `Enum`), a list (`ListValue`), an input object
(`ObjectValue`), or a `Variable`.

Within the scope of the {FieldSelectionMap}, the relationship between input and output is
established by defining the `Value` of the argument as a selection of fields from the output object.

Yet only certain types of `Value` have a semantic meaning. 
`ObjectValue` and `ListValue` are used to define the structure of the value. 
Scalar values, on the other hand, do not carry semantic importance in this context, and variables

> are excluded as they do not exist. 

Reviewer: I wouldn't say variables don't exist, but we may consider them out of scope for now. I think there are legitimate use cases for allowing them though.

Given that these potential values do not align with the standard literals defined in the GraphQL
specification, a new literal called {SelectedValue} is introduced, along with {SelectedObjectValue},

Beyond these literals, an additional literal called {Path} is necessary.

### Name

Is equivalent to the {Name} defined in the 
[GraphQL specification](https://spec.graphql.org/October2021/#Name)

### Path
Path :: 
    - FieldName
    - Path . FieldName
    - FieldName < TypeName > . Path

FieldName ::
    - Name

TypeName ::
    - Name

The {Path} literal is a string used to select a single output value from the _return type_ by
specifying a path to that value. 
This path is defined as a sequence of field names, each separated by a period (`.`) to create 
segments. 

``` example
book.title
```

Each segment specifies a field in the context of the parent, with the root segment referencing a
field in the _return type_ of the query. 

> Arguments are not allowed in a {Path}.

Reviewer 2: Arguments are currently supported for `@requires` in Federation: https://www.apollographql.com/docs/federation/entities-advanced/#using-requires-with-fields-that-take-arguments

For `@is` I agree that arguments do not need to be supported, if we are reusing this Appendix for other directives we could change this to a `MAY be used but are context dependent`

To select a field when dealing with abstract types, the segment selecting the parent field must 
specify the concrete type of the field using angle brackets after the field name if the field is not 
defined on an interface.

> In the following example, the path `mediaById.<Book>.isbn` specifies that `mediaById` returns a

Reviewer 2: ```suggestion
In the following example, the path `mediaById<Book>.isbn` specifies that `mediaById` returns a
```

`Book`, and the `isbn` field is selected from that `Book`.

``` example
mediaById<Book>.isbn
```

### SelectedValue
SelectedValue :: 
    - Path
    - SelectedObjectValue
    - Path . SelectedObjectValue
    - SelectedValue | SelectedValue 

A {SelectedValue} is defined as either a {Path} or a {SelectedObjectValue}

A {Path} is designed to point to only a single value, although it may reference multiple fields
depending on the return type. To allow selection from different paths based on
type, a {Path} can include multiple paths separated by a pipe (`|`).

In the following example, the value could be `title` when referring to a `Book` and `movieTitle`
when referring to a `Movie`. 

``` example
mediaById<Book>.title | mediaById<Movie>.movieTitle
```

The `|` operator can be used to match multiple possible {SelectedValue}. This operator is applied
when mapping an abstract output type to a `@oneOf` input type.

```example

> { movieId: <Movie>.id } | { productId: <Product>.id }

Reviewer 2: Is `<Movie>.id` valid on it's own? Don't we need a `FieldName` first?

```
{ movieId: media<Movie>.id } | { productId: media<Product>.id }
```

Author: maybe we do requrie the `PathSegment` after all:
```
Path :: 
    - < TypeName > . PathSemgent
    - PathSemgent

PathSegment :: 
    - FieldName
    - FieldName . PathSemgent
    - FieldName < TypeName > . PathSemgent

FieldName ::
    - Name

TypeName ::
    - Name
    ```

Reviewer 2: What about ?

```
Path :: 
    - FieldName
    - FieldName < TypeName >
    - Path . Path 

FieldName ::
    - Name

TypeName ::
    - Name
```

Author: i will go with the suggestin above for now

This output type is always a composite type, but the way it's determined can vary depending on the directive and context in which the {FieldSelectionMap} is used.

For example, when used with the `@is` directive, the {FieldSelectionMap} maps between the argument and fields in the return type of the field. 
However, when used with the `@require` directive, it maps between the argument and fields in the object type on which the field is defined.

Consider this example:

```graphql
type Product {
  id: ID!
  delivery(
    zip: String!
    size: Int! @require(field: "dimension.size")
    weight: Int! @require(field: "dimension.weight")
  ): DeliveryEstimates
}
```

In this case, `"dimension.size"` and `"dimension.weight"` refer to fields of the `Product` type, not the `DeliveryEstimates` return type.

Consequently, a {FieldSelectionMap} must be interpreted in the context of a specific argument, its associated directive, and the relevant output type as determined by that directive's behavior.

**Examples**

Scalar fields can be mapped directly to arguments.

This example maps the `Product.weight` field to the `weight` argument:

```graphql example
type Product {

>     shippingCost(weight: Float @require(field: "weight")): Currency

Reviewer 3: just wondering -> with all those examples would it make simpler to grasp if the other types (well their stubs) were present? (this gets somewhat confusing once you get to the complex mapping of top level field and nested fields)

*also those other types are included when talking about the syntax

Author: Well it's a big confusing if we add 'weight' here, as this would represent the already "merged" type from the composed schema instead of a type on a source schema. I think we should leave it out here.

}
```

The path does NOT affect the structure of the input object. It is only used to traverse the output object:

```graphql example
type Product {
    shippingCost(dimension: DimensionInput @require(field: "{ width: size.width height: size.height }")): Currency
}
```

To avoid repeating yourself, you can prefix the selection with a path that ends in a dot to traverse INTO the output type. 

This affects how fields get interpreted but does NOT affect the structure of the input object:

```graphql example
type Product {
    shippingCost(dimension: DimensionInput @require(field: "dimension.{ width height }")): Currency
}
```

This example is equivalent to the previous one:

```graphql example
type Product {
    shippingCost(dimension: DimensionInput @require(field: "size.{ width height }")): Currency
}
```

The path syntax is required for lists because list-valued path expressions would be ambiguous otherwise. 

This example is NOT allowed because it lacks the dot syntax for lists:

```graphql counter-example
type Product {
    shippingCost(dimensions: [DimensionInput] @require(field: "{ width: dimensions.width height: dimensions.height }")): Currency
}
```

> Instead, use the path syntax and angle brackets to specify the list elements: 

Reviewer 4: ```suggestion
Instead, use the path syntax and brackets to specify the list elements: 
```

```example
{ movieId: <Movie>.id } | { productId: <Product>.id }
```

```example
{ nested: { movieId: <Movie>.id } | { productId: <Product>.id }}
```

### SelectedObjectValue
SelectedObjectValue :: 
    - { SelectedObjectField+ }

SelectedObjectField :: 
    - Name: SelectedValue

{SelectedObjectValue} are unordered lists of keyed input values wrapped in curly-braces `{}`.
It has to be used when the expected input type is an object type.

This structure is similar to the `ObjectValue` defined in the GraphQL specification, but it differs by allowing the inclusion of {Path} values within a {SelectedValue}, thus extending the traditional `ObjectValue` capabilities to support direct path selections.

A {SelectedObjectValue} following a {Path} is scoped to the type of the field selected by the {Path}.
This means that the root of all {SelectedValue} inside the selection is no longer scoped to the root (defined by `@is` or `@require`) but to the field selected by the {Path}. The {Path} does not effect the structure of the input type.

This allows to reduce repetition in the selection.

The following example is valid:

```graphql example
type Product {
    dimension: Dimension!

>     shippingCost(dimension: DimensionInput! @require(field: "dimension.{ size weight }")): Int! @lookup

Reviewer 3: i think those `@lookup`s should be dropped?

Author: Done

}
```

Instead, use the path syntax and angle brackets to specify the list elements: 

```graphql example
type Product {
    shippingCost(dimensions: [DimensionInput] @require(field: "dimensions[{ width height }]")): Currency
}
```

> With the path syntax it is possible to also select fields from a list of nested objects

Reviewer 4: ```suggestion
With the path syntax it is possible to also select fields from a list of nested objects:
```

{ nested: { movieId: <Movie>.id } | { productId: <Product>.id }}
```

### SelectedObjectValue
SelectedObjectValue :: 
    - { SelectedObjectField+ }

SelectedObjectField :: 
    - Name: SelectedValue

{SelectedObjectValue} are unordered lists of keyed input values wrapped in curly-braces `{}`.
It has to be used when the expected input type is an object type.

This structure is similar to the `ObjectValue` defined in the GraphQL specification, but it differs by allowing the inclusion of {Path} values within a {SelectedValue}, thus extending the traditional `ObjectValue` capabilities to support direct path selections.

A {SelectedObjectValue} following a {Path} is scoped to the type of the field selected by the {Path}.

> This means that the root of all {SelectedValue} inside the selection is no longer scoped to the root (defined by `@is` or `@require`) but to the field selected by the {Path}. The {Path} does not effect the structure of the input type.

Reviewer 4: ```suggestion
This means that the root of all {SelectedValue} inside the selection is no longer scoped to the root (defined by `@is` or `@require`) but to the field selected by the {Path}. The {Path} does not affect the structure of the input type.
```

> This allows to reduce repetition in the selection.

Reviewer 4: ```suggestion
This allows for reducing repetition in the selection.
```

The following example is valid:

```graphql example
type Product {
    dimension: Dimension!
    shippingCost(dimension: DimensionInput! @require(field: "dimension.{ size weight }")): Int! @lookup
}
```

The following example is equivalent to the previous one:

```graphql example
type Product {
    dimensions: Dimension!
    shippingCost(dimensions: DimensionInput! @require(field: "{ size: dimensions.size weight: dimensions.weight }")): Int! @lookup
}
```

### SelectedListValue
SelectedListValue :: 
    - [ SelectedValue ]

A {SelectedListValue} is an ordered list of {SelectedValue} wrapped in square brackets `[]`. 

> It is used to express semantic equivalence between a an argument expecting a list of values and the values of a list fields within the output object.

Reviewer 4: ```suggestion
It is used to express semantic equivalence between an argument expecting a list of values and the values of a list field within the output object.
```

> The {SelectedListValue} differs from the `ListValue` defined in the GraphQL specification by only allowing one {SelectedValue} as and element.

Reviewer 4: ```suggestion
The {SelectedListValue} differs from the `ListValue` defined in the GraphQL specification by only allowing one {SelectedValue} as an element.
```

The following example is valid:

```graphql example
type Product {
    parts: [Part!]!
    partIds(partIds: [ID!]! @require(field: "parts[id]")): [ID!]! 
}
``` 

In this example, the `partIds` argument is semantically equivalent to the `id` fields of the `parts` list.

The following example is invalid because it uses multiple {SelectedValue} as elements:

```graphql counter-example
type Product {
    parts: [Part!]!
    partIds(parts: [PartInput!]! @require(field: "parts[id name]")): [ID!]! 
}

input PartInput {
    id: ID!
    name: String!
}
``` 

A {SelectedObjectValue} can be used as an element of a {SelectedListValue} to select multiple object fields as long as the input type is a list of structurally equivalent objects.

Similar to {SelectedObjectValue}, a {SelectedListValue} following a {Path} is scoped to the type of the field selected by the {Path}.

> This means that the root of all {SelectedValue} inside the selection is no longer scoped to the root (defined by `@is` or `@require`) but to the field selected by the {Path}. The {Path} does not effect the structure of the input type.

Reviewer 4: ```suggestion
This means that the root of all {SelectedValue} inside the selection is no longer scoped to the root (defined by `@is` or `@require`) but to the field selected by the {Path}. The {Path} does not affect the structure of the input type.
```
